cosmos/lib/cosmos/packets/structure.rb

```diff
@@ -102,24 +102,20 @@ def initialize(default_endianness = BinaryAccessor::HOST_ENDIANNESS, buffer = ''
       def read_item(item, value_type = :RAW, buffer = @buffer)
         return nil if item.data_type == :DERIVED
 
-        if buffer
-          if item.array_size
-            return BinaryAccessor.read_array(item.bit_offset, item.bit_size, item.data_type, item.array_size, buffer, item.endianness)
-          else
-            return BinaryAccessor.read(item.bit_offset, item.bit_size, item.data_type, buffer, item.endianness)
-          end
+        buffer = allocate_buffer() unless buffer
+        if item.array_size
+          return BinaryAccessor.read_array(item.bit_offset, item.bit_size, item.data_type, item.array_size, buffer, item.endianness)
         else
-          raise "No buffer given to read_item"
+          return BinaryAccessor.read(item.bit_offset, item.bit_size, item.data_type, buffer, item.endianness)
         end
       end
 
       # Get the length of the buffer used by the structure
       #
       # @return [Integer] Size of the buffer in bytes
       def length
-        return @buffer.length if @buffer
-
-        return 0
+        allocate_buffer()
```

Comment:
This name is a little misleading and I deleted my previous comment. It's basically allocate buffer if needed. Maybe `get_local_buffer()` (I see you use the term local below).

Reply:
I'll change to allocate_buffer_if_needed